src/schemas/ev42/ev42_rw.cpp

```diff
@@ -200,31 +242,83 @@ HDFWriterModule::reopen(hdf5::node::Group &HDFGroup) {
+      NeXusDataset::Amplitude(AdcGroup, NeXusDataset::Mode::Open);
+  PeakAreaDataset = NeXusDataset::PeakArea(AdcGroup, NeXusDataset::Mode::Open);
+  BackgroundDataset =
+      NeXusDataset::Background(AdcGroup, NeXusDataset::Mode::Open);
+  ThresholdTimeDataset =
+      NeXusDataset::ThresholdTime(AdcGroup, NeXusDataset::Mode::Open);
+  PeakTimeDataset = NeXusDataset::PeakTime(AdcGroup, NeXusDataset::Mode::Open);
+}
 
 void HDFWriterModule::write(FlatbufferMessage const &Message) {
   if (!ds_event_time_offset) {
     throw FileWriter::HDFWriterModuleRegistry::WriterException(
         "Error, time of flight not present.");
   }
-  auto fbuf = get_fbuf(Message.data());
-  auto w1ret = this->ds_event_time_offset->append_data_1d(
-      fbuf->time_of_flight()->data(), fbuf->time_of_flight()->size());
-  auto w2ret = this->ds_event_id->append_data_1d(fbuf->detector_id()->data(),
-                                                 fbuf->detector_id()->size());
-  if (w1ret.ix0 != w2ret.ix0) {
+  auto EventMsgFlatbuffer = GetEventMessage(Message.data());
+  auto EventTimeAppendResultInfo = ds_event_time_offset->append_data_1d(
+      EventMsgFlatbuffer->time_of_flight()->data(),
+      EventMsgFlatbuffer->time_of_flight()->size());
+  auto EventIdAppendResultInfo =
+      ds_event_id->append_data_1d(EventMsgFlatbuffer->detector_id()->data(),
+                                  EventMsgFlatbuffer->detector_id()->size());
+  if (EventTimeAppendResultInfo.ix0 != EventIdAppendResultInfo.ix0) {
     Logger->warn("written data lengths differ");
   }
-  auto pulse_time = fbuf->pulse_time();
-  this->ds_event_time_zero->append_data_1d(&pulse_time, 1);
-  uint32_t event_index = w1ret.ix0;
-  this->ds_event_index->append_data_1d(&event_index, 1);
-  total_written_bytes += w1ret.written_bytes;
+  auto pulse_time = EventMsgFlatbuffer->pulse_time();
+  ds_event_time_zero->append_data_1d(&pulse_time, 1);
+  uint32_t event_index = EventTimeAppendResultInfo.ix0;
+  ds_event_index->append_data_1d(&event_index, 1);
+  total_written_bytes += EventTimeAppendResultInfo.written_bytes;
   ts_max = std::max(pulse_time, ts_max);
   if (total_written_bytes > index_at_bytes + index_every_bytes) {
-    this->ds_cue_timestamp_zero->append_data_1d(&ts_max, 1);
-    this->ds_cue_index->append_data_1d(&event_index, 1);
+    ds_cue_timestamp_zero->append_data_1d(&ts_max, 1);
+    ds_cue_index->append_data_1d(&event_index, 1);
     index_at_bytes = total_written_bytes;
   }
+
+  if (RecordAdcPulseDebugData) {
+    writeAdcPulseData(Message);
+  }
+}
+void HDFWriterModule::writeAdcPulseData(FlatbufferMessage const &Message) {
+  auto EventMsgFlatbuffer = GetEventMessage(Message.data());
+  if (EventMsgFlatbuffer->facility_specific_data_type() !=
+      FacilityData::AdcPulseDebug) {
+    return;
+  }
+  auto AdcPulseDebugMsgFlatbuffer =
+      EventMsgFlatbuffer->facility_specific_data_as_AdcPulseDebug();
+
+  auto AmplitudePtr = AdcPulseDebugMsgFlatbuffer->amplitude()->data();
```

Comment:
We are probably at a point where it is worth it to write a template function in order to minimise copy and paste code. Something similar to the following should probably work:
```
template <typename DataType>
ArrayAdapter<const DataType> getFBVectorAsArrayAdapter(const flatbuffers::Vector<DataType> *Data) {
  return {Data->data(), Data->size()};
}
```

Reply:
:+1: Done